ada_feeding_perception/ada_feeding_perception/segment_all_items.py

```diff
@@ -0,0 +1,1279 @@
+        """
+        self._node.get_logger().info("Running GroundingDINO...")
+
+        # Set the initial time to measure the elapsed time running GroundingDINO on the
+        # desired image and text prompts.
+        inference_time = time.time()
+
+        # Convert image to Image pillow
+        image_pil, image_transformed = self.load_image(image)
+
+        # Lowercase and strip the caption
+        caption = caption.lower().strip()
+
+        # Run GroundingDINO on the image using the input caption
+        image_transformed = image_transformed.to(device=self.device)
+        with torch.no_grad():
+            outputs = self.groundingdino(
+                image_transformed[None],
+                captions=[caption],
+            )
+        logits = outputs["pred_logits"].sigmoid()[0]
+        boxes = outputs["pred_boxes"][0]
+        self._node.get_logger().info("... Done")
+
+        # Filter the output based on the box and text thresholds
+        boxes_cxcywh = {}
+        logits_filt = logits.cpu().clone()
+        boxes_filt = boxes.cpu().clone()
+        filt_thresh_mask = logits_filt.max(dim=1)[0] > box_threshold
+        logits_filt = logits_filt[filt_thresh_mask]
+        boxes_filt = boxes_filt[filt_thresh_mask]
+
+        # Tokenize the caption
+        tokenizer = self.groundingdino.tokenizer
+        caption_tokens = tokenizer(caption)
+
+        # Build the dictionary of bounding boxes for each food item label detected
+        for logit, box in zip(logits_filt, boxes_filt):
+            # Predict phrases based on the bounding boxes and the text threshold
+            phrase = get_phrases_from_posmap(
+                logit > text_threshold, caption_tokens, tokenizer
+            )
+            if phrase not in boxes_cxcywh:
+                boxes_cxcywh[phrase] = []
+            boxes_cxcywh[phrase].append(box.cpu().numpy())
+
+        # Define height and width of image
+        height, width, _ = image.shape
+
+        # Convert the bounding boxes outputted by GroundingDINO to the following format
+        # [top left x-value, top left y-value, bottom right x-value, bottom right y-value]
+        # and unnormalize the bounding box coordinate values
+        boxes_xyxy = {}
+        for phrase, boxes in boxes_cxcywh.items():
+            boxes_xyxy[phrase] = []
+            for box in boxes:
+                # Scale the box from percentage values to pixel values
+                box = np.multiply(box, np.array([width, height, width, height]))
+                center_x, center_y, w, h = box
+                # Get the bottom left and top right coordinates of the box
+                x0 = center_x - (w / 2)
+                y0 = center_y - (h / 2)
+                x1 = x0 + w
+                y1 = y0 + h
+                boxes_xyxy[phrase].append([x0, y0, x1, y1])
+
+        # Measure the elapsed time running GroundingDINO on the image prompt
+        inference_time = int(round((time.time() - inference_time) * 1000))
```

Comment:
Is this meant to be logged?

Reply:
Also, why are you multiplying by 1000?

In general, I find it more readable to specify the units of measurements after the variable name itself, e.g., in this case maybe it should be `inference_time_ms` given the factor of `1000`?